pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -456,3 +487,315 @@ func (plugin *kubenetNetworkPlugin) getNsenterPath() (string, error) {
+		return
+	}
+
+	// Get iptables-save output so we can check for existing chains and rules.
+	// This will be a map of chain name to chain with rules as stored in iptables-save/iptables-restore
+	existingNATChains := make(map[utiliptables.Chain]string)
+	iptablesSaveRaw, err := plugin.iptables.Save(utiliptables.TableNAT)
+	if err != nil { // if we failed to get any rules
+		glog.Errorf("Failed to execute iptables-save, syncing all rules: %v", err)
+	} else { // otherwise parse the output
+		existingNATChains = utiliptables.GetChainLines(utiliptables.TableNAT, iptablesSaveRaw)
+	}
+
+	natChains := bytes.NewBuffer(nil)
+	natRules := bytes.NewBuffer(nil)
+	writeLine(natChains, "*nat")
+	// Make sure we keep stats for the top-level chains, if they existed
+	// (which most should have because we created them above).
+	if chain, ok := existingNATChains[kubenetHostportChain]; ok {
+		writeLine(natChains, chain)
+	} else {
+		writeLine(natChains, utiliptables.MakeChainLine(kubenetHostportChain))
+	}
+	if chain, ok := existingNATChains[kubenetPostroutingChain]; ok {
+		writeLine(natChains, chain)
+	} else {
+		writeLine(natChains, utiliptables.MakeChainLine(kubenetPostroutingChain))
+	}
+	if chain, ok := existingNATChains[kubenetMarkChain]; ok {
+		writeLine(natChains, chain)
+	} else {
+		writeLine(natChains, utiliptables.MakeChainLine(kubenetMarkChain))
+	}
+
+	// Install the kubenet-specific postrouting rules. We use a whole chain for
+	// this so that it is easier to flush and change, for example if the mark
+	// value should ever change.
+	writeLine(natRules, []string{
+		"-A", string(kubenetPostroutingChain),
+		"-m", "comment", "--comment", `"kubenet hostport traffic requiring SNAT"`,
+		"-m", "mark", "--mark", masqueradeMark,
+		"-j", "MASQUERADE",
+	}...)
+
+	// Install the kubenet-specific masquerade mark rule. We use a whole chain for
+	// this so that it is easier to flush and change, for example if the mark
+	// value should ever change.
+	writeLine(natRules, []string{
+		"-A", string(kubenetMarkChain),
+		"-j", "MARK", "--set-xmark", masqueradeMark,
+	}...)
+
+	// Accumulate NAT chains to keep.
+	activeNATChains := map[utiliptables.Chain]bool{} // use a map as a set
+
+	for containerPort, target := range containerPortMap {
+		protocol := strings.ToLower(string(containerPort.Protocol))
+		hostportChain := hostportChainName(containerPort, target.podFullName)
+		if chain, ok := existingNATChains[hostportChain]; ok {
+			writeLine(natChains, chain)
+		} else {
+			writeLine(natChains, utiliptables.MakeChainLine(hostportChain))
+		}
+
+		activeNATChains[hostportChain] = true
+
+		// Capture the hostport
+		args := []string{
+			"-A", string(kubenetHostportChain),
+			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, target.podFullName, containerPort.HostPort),
+			"-m", protocol, "-p", protocol,
+			"--dport", fmt.Sprintf("%d", containerPort.HostPort),
+		}
+
+		writeLine(natRules, append(args, "-j", string(kubenetMarkChain))...)
```

Comment:
Why do we need to MASQUERADE hostport traffic?  That is pretty hostile to anyone who wants to use it.